### PR TITLE
Adds support of Hadoop compression codec.

### PR DESCRIPTION
**Pull request by cheqianh**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jobarr-amzn — APPROVED** [review]
We could stand to add a little commentary to our record reader about why we maintain the `Seekable`/`InputStream` distinction, and why we reference each one when we do (e.g. in `getProgress()`, `getPos()`, and `close()`).

**Review by jobarr-amzn — COMMENTED** [review]
We could also use a comment on `getProgress()`, explaining why the `start == end` case results in `getProgress() == 0`. (We're following precedence in Hadoop's TextInputFormat). 

TextInputFormat doesn't say *why* it does this, nor does the RecordReader documentation have any specific ruling on how `getProgress()` ought to relate with the underlying machinery. This is [cargo cult programming](https://en.wikipedia.org/wiki/Cargo_cult_programming) - we should at least be honest about it. 

In fact I'd encourage you to dig a bit farther and look for other implementations of `getProgress()` as well as the history of `TextInputFormat.getProgress()` to see if you can learn why it does this. Is it a shortcut because `x - x == 0`? But why should a zero-size (empty?) split be marked as zero progress? Does it make more sense to rather say that it is *done* and that progress is `1.0`?

This is not core functionality, but still we can do better than a blind copy here.

**Comment by cheqianh** [comment]
Sure I think it's a good idea to add more comments. 

It's also a good opportunity to look into other serde packages and figure out how/when the APIs are called. Will work on it and get back to you soon.

**Comment by cheqianh** [comment]
Discussed offline, Added some comments. tested locally it worked well for gzip.

File: serde/src/main/java/com/amazon/ionhiveserde/formats/IonInputFormat.java [hunk 1]
```diff
@@ -116,19 +135,25 @@ public BytesWritable createValue() {
...
 
         @Override
         public final void close() throws IOException {
-            this.fsDataInputStream.close();
+            this.in.close();
         }
 
         @Override
+        // We don't know why `progress` is equal to 0 instead of 1 when start == end, what Hadoop is doing since long
+        // times ago is setting it to 0.
```
Comment (jobarr-amzn) [inline serde/src/main/java/com/amazon/ionhiveserde/formats/IonInputFormat.java:151]: Just for fun let's link to the source of these behaviors.

The original split support issue: https://issues.apache.org/jira/browse/HADOOP-451
The original patch that introduced this:  https://svn.apache.org/viewvc/lucene/hadoop/trunk/src/java/org/apache/hadoop/mapred/TextInputFormat.java?r1=469596&r2=488438&pathrev=502021&diff_format=h

**Review by jobarr-amzn — APPROVED** [review]

**Comment by cheqianh** [comment]
Added links to the source. I'll do an overall testing again before sending this PR to the official repo.

**Pull request merged**